enhancements/console/dynamic-plugins.md

```diff
@@ -0,0 +1,374 @@
+to load plugins at runtime. Module federation allows a JavaScript application
+to dynamically load code from another application while sharing dependencies.
+It also allows console to share its components with plugins built and bundled
+separately.
+
+Plugins will need to be built with Webpack 5 to use module federation.
+
+We have a pull request with a [working prototype](https://github.com/openshift/console/pull/6101)
+that uses module federation.
+
+### Shared Dependencies
+
+Console will expose React and Patternfly 4 as shared dependencies to plugins.
+Plugins can use any Patternfly component. Only a single version of React and
+Patternfly will be loaded.
+
+Plugins will not be able to use legacy Patternfly 3 components.
+
+### Delivering Plugins
+
+Plugins are delivered through operators. The operators will typically create a
+deployment on the platform with an HTTP server to host the plugin and expose it
+using a Kubernetes `Service`. The `Service` must use HTTPS and a
+[service serving certificate](https://docs.openshift.com/container-platform/4.4/security/certificates/service-serving-certificate.html).
+The console backend will load the plugin directly from the `Service` using the
+service CA bundle.
+
+Operators declare that they have a console plugin available in the
+`ClusterServiceVersion` resource. This can be done through the
+`console.openshift.io/plugins` annotation on the CSV. This annotation is only a
+hint to console that a dynamic plugin is available. Console will not load the
+plugin until a cluster admin enables it, and each plugin is enabled or disabled
+individually. The user must have write access to the console operator config to
+enable or disable plugins.
+
+The value is a serialized JSON array of objects. Each object has a service
+name, port, and path to the plugin manifest.
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  annotations:
+    console.openshift.io/plugins: '[{ "service": "my-console-plugin", "port": 8443, "manifest": "/manifest.json" }]'
+```
+
+The namespace for the `Service` is assumed to be the same namespace as the
+`ClusterServiceVersion`. In the case of global operators, it will always be the
+namespace of the global `ClusterServiceVersion` (not a CSV copy).
+
+Plugins will be disabled by default. They will need to be manually enabled by a
+cluster administrator before console will load any plugin code. Console will
+provide a UI for enabling plugins. The list of enabled plugins is set in the
+console operator config.
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: Console
+metadata:
+  name: cluster
+spec:
+  managementState: Managed
+  plugins:
+  - service: my-console-plugin
+    namespace: openshift-operators
+    port: 8443
+    manifest: /manifest.json
+```
+
+This will trigger a new rollout of the console `Deployment` with the updated
+plugins. Note that this does mean the user will need to refresh the browser
+to see the new plugins.
```

Comment:
I guess this is sometimes a good thing, because you won't be auto-updated to a new deployment where a plugin you're currently using has been removed.  Or maybe that's not a good thing?  But having local JavaScript and other assets in your browser doesn't mean that the cluster still supports whatever backing API endpoints the plugin is driving.  Seems like we need something to at least wave a notification "your view is stale; click *here* to update" or whatever.  Console could poll some sort of deployment-identifier API to detect the staleness.